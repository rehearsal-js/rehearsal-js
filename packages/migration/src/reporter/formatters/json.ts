import { Report } from '../reporter';

export default function json(report: Report): string {
  return JSON.stringify(report, null, 2);
}
