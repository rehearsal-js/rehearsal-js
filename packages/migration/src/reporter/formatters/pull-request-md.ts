import { Report } from '../types';

export default function pullRequestMd(report: Report): string {
  const fileNames = [...new Set(report.items.map((item) => item.file))];

  let text = ``;

  text += `### Summary:\n`;
  text += `Typescript Version: ${report.summary.tsVersion}\n`;
  text += `Files Tested: ${fileNames.length}\n`;
  text += `\n`;
  text += `### Results:\n`;

  for (const fileName of fileNames) {
    const items = report.items.filter((item) => item.file === fileName);
    const relativeFileName = fileName.replace(report.summary.basePath, '');

    text += `\n`;
    text += `#### File: ${relativeFileName}, issues: ${items.length}:\n`;

    for (const item of items) {
      text += `\n`;
      text += `**${item.category} TS${item.code}**: ${
        item.fixed ? 'FIXED' : 'NEED TO BE FIXED MANUALLY'
      }\n`;
      text += `${item.hint}\n`;
      text += `Code: \`${item.nodeText}\`\n`;
    }
  }

  return text;
}
