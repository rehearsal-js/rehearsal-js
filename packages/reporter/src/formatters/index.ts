export * from './json-formatter.js';
export * from './md-formatter.js';
export * from './sonarqube-formatter.js';
export * from './sarif-formatter.js';
