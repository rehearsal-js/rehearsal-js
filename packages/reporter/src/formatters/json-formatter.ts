import { Report } from '../types';

export function jsonFormatter(report: Report): string {
  return JSON.stringify(report, null, 2);
}
