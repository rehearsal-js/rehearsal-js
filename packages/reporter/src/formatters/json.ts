import { Report } from '../types';

export function json(report: Report): string {
  return JSON.stringify(report, null, 2);
}
