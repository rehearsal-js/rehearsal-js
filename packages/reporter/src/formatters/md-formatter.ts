import type { Report, FormatterBase } from '../types.js';

export class MarkdownFormatter implements FormatterBase {
  static extension = '.md';
  static getReport(report: Report): string {
    const fileNames = [...new Set(report.items.map((item) => item.analysisTarget))];

    let text = ``;

    text += `### Summary:\n`;

    for (const block of report.summary) {
      text += `Project Name: ${block.projectName}\n`;
      text += `Typescript Version: ${block.tsVersion}\n`;
      text += `Report path: ${block.reportOutDir}\n`;
      text += `timestamp: ${block.timestamp}\n`;
      text += `\n`;
    }

    text += `### Results:\n`;

    for (const fileName of fileNames) {
      const items = report.items.filter((item) => item.analysisTarget === fileName);

      text += `\n`;
      text += `#### File: ${fileName}, issues: ${items.length}:\n`;

      for (const item of items) {
        text += `\n`;
        text += `**${item.category} ${item.ruleId}**: 'REVIEW REQUIRED'\n`;
        text += `${item.hint}\n`;
        text += `Code: \`${item.nodeText}\`\n`;
      }
    }

    return text;
  }
}
