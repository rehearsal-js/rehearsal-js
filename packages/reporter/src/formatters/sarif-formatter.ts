import { Log, Run, ReportingDescriptor, Result, Artifact } from 'sarif';
import { Report, ReportItem } from '../types';

function initReport(): Log{
  return {
    version: "2.1.0" as "2.1.0",
    $schema: "http://json.schemastore.org/sarif-2.1.0-rtm.4",
    runs: [],
  }
}

function initRun(): Run {
  return {
    tool: {
      driver: {
        name: '@rehearsal/migrate',
        informationUri: 'https://github.com/rehearsal-js/rehearsal-js',
        rules: []
      }
    },
    artifacts: [],
    results: []
  }
}

function buildRule(item: ReportItem): ReportingDescriptor{
  return {
    id: 'TS' + item.code,
    shortDescription: {
      text: item.message
    },
    // helpUri: item.helpUri
  }
}

function buildArtifact(item: ReportItem, existingArtifacts: Artifact[]) {
  console.log('artifacts-------------------------', existingArtifacts);
  console.log('item-------------------', item);
  let processedFiles: Artifact[] = [];
  const analysisTarget = {
    location: {
      uri: item.analysisTarget
    },
    roles: ['analysisTarget'],
    properties: {
      codemod: item.fixedFiles
    }
  }
  item.files.forEach((file) => {
    const existingArtifact = existingArtifacts.find((artifact) => 
      artifact.location!.uri === file.fileName
    );
    console.log('existingArtifact', existingArtifact);

    if (existingArtifact){
      if(!existingArtifact.roles!.includes(file.role)) {
        existingArtifact.roles = [...existingArtifact.roles!, file.role];
      }
      if(existingArtifact.properties!.codemod === false && file.codemod === true) {
        existingArtifact.properties!.codemod = true;
      }
      if(existingArtifact.properties!.commentAdded === false && file.commentAdded === true) {
        existingArtifact.properties!.commentAdded = true;
      }
    } else {
      processedFiles.push({
        location: {
          uri: file.fileName
        },
        roles: [ file.role ],
        properties: {
          codemod: file.codemod,
          commentAdded: file.commentAdded,
        }
      })
    }
  });
  return processedFiles;
}

function levelConverter(category: string): Result.level {
  switch(category) {
    case 'Warning': 
    return 'warning';
    case "Error":
      return 'error';
    case "Suggestion":
      return 'note';
    case "Message":
      return 'note';
    default:
      return 'none';
  }
}

function buildResult(item: ReportItem, ruleIndex: number) {
  const primaryFile = item.files.find((file) => file.role === 'analysisTarget');
  const locations = item.files.map((file) => ({
    physicalLocation: {
      artifactLocation: {
        uri: file.fileName
      },
      region: {
        startLine: item.location.line,
        startColumn:item.location.start, 
      }
    }
  }))
  return {
    ruleId: 'TS' + item.code,
    ruleIndex,
    level: levelConverter(item.category),
    message: {
      text: item.hint,
    },
    analysisTarget: {
      uri: primaryFile?.fileName
    },
    // kind: 'fail' as 'fail', //'notApplicable'| 'pass' | 'fail' | 'review' | 'open' | 'informational'
    locations
      // {
      //   physicalLocation: {
      //     artifactLocation: {
      //       uri: item.file
      //     },
      //     region: {
      //       startLine: item.location.line,
      //       startColumn: item.location.start
      //     }
      //   },
      //   relationships: [
      //     {
      //       kinds: ['importedFrom'],
      //       target: 1,
      //     }
      //   ]
      // },
  }
}

function createRun(items: ReportItem[]): Run{
  let run = initRun();
  for(const item of items) {
    run.tool.driver.rules!.push(buildRule(item));
    const artifact = buildArtifact(item, run.artifacts!);
    if(artifact) {
      run.artifacts! = [...run.artifacts!, ...artifact ]
    }
    // artifact && run.artifacts!.concat(artifact);
    run.results!.push(buildResult(item, run.tool.driver.rules!.length - 1))
  }
  return run;

}
export function sarifFormatter(data: Report): string {
  let report = initReport();
  if(!data || !data.items) {
    return JSON.stringify(report, null, 2);
  }
  const run = createRun(data.items);
  report.runs.push(run);
  return JSON.stringify(report, null, 2);
}