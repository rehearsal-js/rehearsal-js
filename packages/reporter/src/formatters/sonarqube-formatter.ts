import { isAbsolute, resolve } from 'path';
import { Log, PhysicalLocation, Result } from 'sarif';

interface FormatterOptions {
  cwd: string;
  outputFile?: string;
}

const SONARQUBE_SEVERITY: Record<Result.kind, string> = {
  notApplicable: 'INFO',
  pass: 'INFO',
  fail: 'BLOCKER',
  review: 'CRITICAL',
  open: 'MINOR',
  informational: 'INFO',
};
type ErrorLevel = Extract<Result.level, 'error' | 'warning' | 'note'>;
const SONARQUBE_TYPE: Record<ErrorLevel, string> = {
  note: 'CODE_SMELL',
  warning: 'CODE_SMELL',
  error: 'BUG',
};

function getPhysicalLocation(result: Result): PhysicalLocation | undefined {
  return (result.locations && result.locations[0].physicalLocation) ?? undefined;
}

function getFilePath(physicalLocation: PhysicalLocation): string {
  return physicalLocation?.artifactLocation?.uri ?? '';
}

export function sonarqubeFormatter(log: Log, options: FormatterOptions): string | undefined {
  const issues = [];

  const run = log.runs[0];
  const results = run.results || [];

  if (results.length > 0) {
    for (const result of results) {
      const physicalLocation = getPhysicalLocation(result);
      const filePath = physicalLocation ? getFilePath(physicalLocation) : '';
      const absolutePath = isAbsolute(filePath) ? filePath : resolve(options.cwd, filePath);

      issues.push({
        engineId: 'rehearsal-ts',
        ruleId: result.ruleId,
        severity: SONARQUBE_SEVERITY[result.kind ?? 'notApplicable'],
        type: SONARQUBE_TYPE[result.level as ErrorLevel],
        primaryLocation: {
          message: result.message.text ?? '',
          filePath: absolutePath,
          textRange: {
            startLine: physicalLocation?.region?.startLine ?? 0,
            startColumn: physicalLocation?.region?.startColumn ?? 0,
            endLine: physicalLocation?.region?.endLine ?? 0,
            endColumn: physicalLocation?.region?.endColumn ?? 0,
          },
        },
      });
    }
    return JSON.stringify({ issues }, null, 2);
  }
}
