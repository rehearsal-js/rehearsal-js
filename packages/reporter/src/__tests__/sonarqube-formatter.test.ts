import { resolve } from 'path';
import { readJSONSync } from 'fs-extra';
import { describe, expect, test } from 'vitest';
import { sonarqubeFormatter } from '../formatters/sonarqube-formatter';
import type { Report } from '../types';

const rehearsalReportJSON: Report = readJSONSync(
  resolve('./test/fixtures/reporter/rehearsal-report.json')
);

describe('Test sonarqueFormatter', () => {
  test('should transform all fields correctly, irregardless of undefined, missing values', () => {
    const sonarqube = sonarqubeFormatter(rehearsalReportJSON);
    expect(sonarqube).matchSnapshot();
  });
});
