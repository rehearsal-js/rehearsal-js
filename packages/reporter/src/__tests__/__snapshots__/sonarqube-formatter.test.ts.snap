// Vitest Snapshot v1

exports[`Test sonarqueFormatter > should transform all fields correctly, irregardless of undefined, missing values 1`] = `
"{
  \\"issues\\": [
    {
      \\"engineId\\": \\"rehearsal-ts\\",
      \\"ruleId\\": \\"TS6133\\",
      \\"severity\\": \\"BLOCKER\\",
      \\"type\\": \\"BUG\\",
      \\"primaryLocation\\": {
        \\"message\\": \\"The declaration 'react' is never read or used. Remove the declaration or use it.\\",
        \\"filePath\\": \\"/base/path/file1.ts\\",
        \\"textRange\\": {
          \\"startLine\\": 1,
          \\"startColumn\\": 1,
          \\"endLine\\": 1,
          \\"endColumn\\": 1
        }
      }
    },
    {
      \\"engineId\\": \\"rehearsal-ts\\",
      \\"ruleId\\": \\"TS2322\\",
      \\"severity\\": \\"BLOCKER\\",
      \\"type\\": \\"BUG\\",
      \\"primaryLocation\\": {
        \\"message\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
        \\"filePath\\": \\"/base/path/file1.ts\\",
        \\"textRange\\": {
          \\"startLine\\": 1,
          \\"startColumn\\": 10,
          \\"endLine\\": 1,
          \\"endColumn\\": 20
        }
      }
    },
    {
      \\"engineId\\": \\"rehearsal-ts\\",
      \\"ruleId\\": \\"TS2345\\",
      \\"severity\\": \\"BLOCKER\\",
      \\"type\\": \\"BUG\\",
      \\"primaryLocation\\": {
        \\"message\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
        \\"filePath\\": \\"/base/path/file1.ts\\",
        \\"textRange\\": {
          \\"startLine\\": 1,
          \\"startColumn\\": 10,
          \\"endLine\\": 1,
          \\"endColumn\\": 26
        }
      }
    }
  ]
}"
`;
